### DISCUSSION TITLE
PETSC error using Docker

### DISCUSSION THREAD

**Discussion by huisophiewang** [firedrakeproject/firedrake#3912] [Firedrake support]
Hi,

I installed firedrake through virtualenv before, and now I need to switch to docker in order to run my algorithm in the cluster. 
After running "docker run -it firedrakeproject/firedrake bash", and "source activate", I tried to import firedrake, and here's the error I got. I’m not entirely sure what this means. Could someone kindly provide some guidance on how to solve this problem?

(firedrake) firedrake@e67f9aae2552:~/firedrake/bin$ python3 
Python 3.12.3 (main, Nov  6 2024, 18:32:19) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import firedrake
[0]PETSC ERROR: ------------------------------------------------------------------------
[0]PETSC ERROR: Caught signal number 4 Illegal instruction: Likely due to memory corruption
[0]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
[0]PETSC ERROR: or see https://petsc.org/release/faq/#valgrind and https://petsc.org/release/faq/
[0]PETSC ERROR: configure using --with-debugging=yes, recompile, link, and run 
[0]PETSC ERROR: to get more information on the crash.
application called MPI_Abort(MPI_COMM_WORLD, 59) - process 0
[unset]: PMIU_write error; fd=-1 buf=:cmd=abort exitcode=59 message=application called MPI_Abort(MPI_COMM_WORLD, 59) - process 0
:
system msg for write_line failure : Bad file descriptor
(firedrake) firedrake@e67f9aae2552:~/firedrake/bin$

**Comment by Ig-dolci** [comment]
Can you please check if the docker image from `docker pull firedrakeproject/firedrake-vanilla` gives you the same error?

**Reply by huisophiewang** [reply]
There are no errors when I use the firedrake-vanilla docker image. Should I use this one instead?

**Reply by huisophiewang** [reply]
My Python program actually uses PETSc and SLEPc.  I'm using firedrake-vanilla as the base image. When I add this to the dockerfile "RUN bash -c "source firedrake/bin/activate; firedrake-update --slepc", it has the same error again. Any suggestions?

124.1 Installing slepc4py/
124.1 Copying slepc4py to virtualenv to install
174.3 /home/firedrake/firedrake/lib/python3.12/site-packages/pytools/persistent_dict.py:52: RecommendedHashNotFoundWarning: Unable to import recommended hash 'siphash24.siphash13', falling back to 'hashlib.sha256'. Run 'python3 -m pip install siphash24' to install the recommended hash.
174.3   warn("Unable to import recommended hash 'siphash24.siphash13', "
174.3 [0]PETSC ERROR: ------------------------------------------------------------------------
174.3 [0]PETSC ERROR: Caught signal
174.3 [0]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
174.3 [0]PETSC ERROR: or see https://petsc.org/release/faq/#valgrind and https://petsc.org/release/faq/
174.3 [0]PETSC ERROR: configure using --with-debugging=yes, recompile, link, and run 
174.3 [0]PETSC ERROR: to get more information on the crash.
174.3 application called MPI_Abort(MPI_COMM_WORLD, 59) - process 0
174.3 [unset]: PMIU_write error; fd=-1 buf=:cmd=abort exitcode=59 message=application called MPI_Abort(MPI_COMM_WORLD, 59) - process 0
174.3 :
174.3 system msg for write_line failure : Bad file descriptor
------

**Reply by huisophiewang** [reply]
Also, this might be relevant. I'm using Mac.

**Reply by Ig-dolci** [reply]
> There are no errors when I use the firedrake-vanilla docker image. Should I use this one instead?

It seems that one of the packages installed in the Firedrake image might have an architecture issue. Could you try the following to troubleshoot the error?

Inside the `firedrake` image, run:  
 ```bash
 pip uninstall netgen
 ```  
Then, try importing Firedrake again to verify if the error persists. I already saw a similar fail associated with Netgen.

**Reply by Ig-dolci** [reply]
> My Python program actually uses PETSc and SLEPc. I'm using firedrake-vanilla as the base image. When I add this to the dockerfile "RUN bash -c "source firedrake/bin/activate; firedrake-update --slepc", it has the same error again. Any suggestions?
> 
> ## 124.1 Installing slepc4py/
> 124.1 Copying slepc4py to virtualenv to install
> 174.3 /home/firedrake/firedrake/lib/python3.12/site-packages/pytools/persistent_dict.py:52: RecommendedHashNotFoundWarning: Unable to import recommended hash 'siphash24.siphash13', falling back to 'hashlib.sha256'. Run 'python3 -m pip install siphash24' to install the recommended hash.
> 174.3 warn("Unable to import recommended hash 'siphash24.siphash13', "
> 174.3 [0]PETSC ERROR: ------------------------------------------------------------------------
> 174.3 [0]PETSC ERROR: Caught signal
> 174.3 [0]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
> 174.3 [0]PETSC ERROR: or see https://petsc.org/release/faq/#valgrind and https://petsc.org/release/faq/
> 174.3 [0]PETSC ERROR: configure using --with-debugging=yes, recompile, link, and run
> 174.3 [0]PETSC ERROR: to get more information on the crash.
> 174.3 application called MPI_Abort(MPI_COMM_WORLD, 59) - process 0
> 174.3 [unset]: PMIU_write error; fd=-1 buf=:cmd=abort exitcode=59 message=application called MPI_Abort(MPI_COMM_WORLD, 59) - process 0
> 174.3 :
> 174.3 system msg for write_line failure : Bad file descriptor

If you want to have SLEPc working on top of the `firedrake_vanilla` image, you can follow these steps that are working here on my Mac.

1. **Start a Container Using `firedrake_vanilla`**:  
   Run the following command:  
   ```bash
   docker run -it <image_id_from_firedrake_vanilla> /bin/bash
   ```

2. **Activate Firedrake and Install SLEPc**:  
   Inside the container, execute:
   ```bash
   source firedrake/bin/activate
   firedrake-update --slepc
   ```

3. **Save the Updated Container as a New Image**:  
   Exit the container, and from your terminal, commit the changes to create a new Docker image:
   ```bash
   docker commit <container_id> firedrake_slepc:v1
   ```
Now, you can use the new image `firedrake_slepc:v1`, which has SLEPc installed.

**Reply by connorjward** [reply]
I *think* that our Docker containers do not really work on Macs because we only have x86 images, not ARM ones. We should try and fix that.

**Reply by huisophiewang** [reply]
I see. I'll switch to Linux for now. Would the latest Ubuntu work well with firedrake or do you have other recommendations?

**Reply by connorjward** [reply]
Ubuntu will definitely work, but you could also try and regular installation of Firedrake on macOS without Docker.
